### ISSUE TITLE
WildcardPattern.Escape() does not escape lone backticks correctly

### ISSUE THREAD

**Issue by IISResetMe** [PowerShell/PowerShell#16306]
### Prerequisites

- [X] Write a descriptive title.
- [X] Make sure you are able to repro it on the [latest released version](https://github.com/PowerShell/PowerShell/releases)
- [X] Search the existing issues.
- [X] Refer to the [FAQ](https://github.com/PowerShell/PowerShell/blob/master/docs/FAQ.md).
- [X] Refer to [Differences between Windows PowerShell 5.1 and PowerShell](https://docs.microsoft.com/powershell/scripting/whats-new/differences-from-windows-powershell).

### Steps to reproduce

WildcardPattern.Escape() does not escape `` ` ``, and as a result it's unreliable for escaping input for use in wildcard comparisons:

Define a string literal containing a single backtick character:

    $string = 'abc`def'

Create an escaped copy of the string value:

    $escaped = [WildcardPattern]::Escape($string)

Perform wildcard comparison and observe that it does not yield `$true` as expected:

    $string -like $escaped

(That is: `[WildcardPattern]::Escape($string)` should produce a string value equivalent to ```'abc``def'```, not return it as-is)

### Expected behavior

```console
PS> $string -like $escaped
True
```

### Actual behavior

```console
PS> $string -like $escaped
False
```

### Environment data

```powershell
Name                           Value
----                           -----
PSVersion                      7.1.4
PSEdition                      Core
GitCommitId                    7.1.4
OS                             Microsoft Windows 10.0.19042
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

**Comment by IISResetMe** [comment]
@daxian-dbw do we have any good reason to not move ahead and fix `Escape()` here? Happy to do the PR

**Comment by ArmaanMcleod** [comment]
> /cc [@ArmaanMcleod](https://github.com/ArmaanMcleod) If you have an interest to fix.

All good I made a quick PR. It looks like `IsWildcardChar` was left behind and is just missing backtick discrepancy. It should probably just be using `SpecialChars` which has all 5 special characters and is used elsewhere in the code.

**Comment by iSazonov** [comment]
Looking example in OP, we have two single-quoted string. All chars in such strings are literals (no conversions like `n -> CR). So really we should do literal comparison of the strings. `-eq` does this exactly but `-like` doesn't.
The code https://github.com/PowerShell/PowerShell/blob/4e7942135f998ab40fd3ae298b020e161a76d4ef/src/System.Management.Automation/engine/regex.cs#L176-L182

I think we should remove backtick from `SpecialChars` - only wildcards must be there. As result the code will do string comparison (line 180).
We could try this in PR and see test results to get understanding about if it breaks something.

**Comment by ArmaanMcleod** [comment]
So, in the escape method which basic rebuilds a span and checks wildcard chars: https://github.com/PowerShell/PowerShell/blob/4e7942135f998ab40fd3ae298b020e161a76d4ef/src/System.Management.Automation/engine/regex.cs#L233-L249

I tried to fix this to also check for backtick `IsWildcardChar(ch) || ch == escapeChar` which seems to not break anything, but I guess it is also valid what @iSazonov has shown above. 

If a Wildcard pattern is just supposed to escape valid wildcard special chars, then this might not be the fix. 

It's also a bit confusing to understand in the code what is the difference between special chars and wildcard chars, with the former just including a backtick but it seems there was a deliberate choice to keep it out of the wildcard char method: 
https://github.com/PowerShell/PowerShell/blob/4e7942135f998ab40fd3ae298b020e161a76d4ef/src/System.Management.Automation/engine/regex.cs#L438-L441

**Comment by ArmaanMcleod** [comment]
I think @iSazonov and myself have looked into it a bit more and it should be as simple as remove backtick from SpecialChars. With recent PR change it seems to be working.

@IISResetMe When you get time would appreciate a review, this one was a funny one 😄.

**Comment by ArmaanMcleod** [comment]
I've realised this as well seems to break for Argument completers which rely on `WildcardPattern.IsMatch` in C#. Specifically for completers which want to complete CRLF.

XUnit Repro:

```csharp
[Theory]
[InlineData("`r*", "`r`n", true)]
public void TestIsMatch_String(string pattern, string input, bool result)
{
    Assert.Equal(WildcardPattern.Get(pattern, WildcardOptions.IgnoreCase).IsMatch(input), result);
}
```

Result:

```text
[xUnit.net 00:00:00.70]     PSTests.Parallel.WildcardPatternTests.TestIsMatch_String(pattern: "`r*", input: "`r`n", result: True) [FAIL]
[xUnit.net 00:00:00.70]       Assert.Equal() Failure: Values differ
[xUnit.net 00:00:00.70]       Expected: True
[xUnit.net 00:00:00.70]       Actual:   False
```

If you remove backtick from `SpecialChars` this works as expected.

PowerShell on the other hand does not seem to be broken:

```pwsh
> [WildcardPattern]::Get("`r*", [System.Management.Automation.WildcardOptions]::IgnoreCase).IsMatch("`r`n")
True
```

So I am confused 😄.

**Comment by iSazonov** [comment]
> So I am confused 😄.

Single quota vs double quota?

**Comment by ArmaanMcleod** [comment]
Coming back to this one, I believe just escaping directly in the `WildardPattern.Escape()` method was the right solution: https://github.com/ArmaanMcleod/PowerShell/commit/8b98336fab31dbf8017d01b2371f4eaf796c18d8

With some Pester tests:

```pwsh
It "Should be <result> for backtick comparison <lhs> <operator> <rhs>" -TestCases @(
    @{ lhs = 'abc`def'; operator = '-like'; rhs = 'abc`def'; result = $false }
    @{ lhs = 'abc`def'; operator = '-like'; rhs = 'abc``def'; result = $true }
    @{ lhs = 'abc`def'; operator = '-like'; rhs = 'abc````def'; result = $false }
    @{ lhs = 'abc``def'; operator = '-like'; rhs = 'abc````def'; result = $true }
    @{ lhs = 'abc`def'; operator = '-like'; rhs = [WildcardPattern]::Escape('abc`def'); result = $true }
 ) {
    param($lhs, $operator, $rhs, $result)
    $expression = "'$lhs' $operator '$rhs'"
    Invoke-Expression $expression | Should -Be $result
}
```

And results:

```pwsh
[+] Should be $false for backtick comparison abc`def -like abc`def 3ms
[+] Should be $true for backtick comparison abc`def -like abc``def 1ms
[+] Should be $false for backtick comparison abc`def -like abc````def 2ms
[+] Should be $true for backtick comparison abc``def -like abc````def 2ms
[+] Should be $true for backtick comparison abc`def -like abc``def 2ms
```

**Comment by iSazonov** [comment]
> @{ lhs = [WildcardPattern]::Escape('abc`def'); operator = '-eq'; rhs = 'abc``def'; result = $true }

I wonder how it is true.

**Comment by ArmaanMcleod** [comment]
@iSazonov `[WildcardPattern]::Escape('abc`def')` would just turn into `abc``def` after escaping and then its just a direct `-eq` comparison which would return true.

Currently this doesn't work since `WildcardPattern.Escape` is not escaping backticks so it returns false.

**Comment by iSazonov** [comment]
But right side is WildcardPattern where backtick is special char and effective right string will be 'abc`def' vs effective left string 'abc``def'.

**Comment by iSazonov** [comment]
> > But right side is WildcardPattern
> 
> No, -eq will treat it as a normal string literal with backticks in it

So such tests useless in this context. They could be separate, although I don't see much point in that.

**Comment by ArmaanMcleod** [comment]
In that case it's it might be better just keeping the `-like` tests in Pester and the Escape/UnEscape tests in XUnit which test the conversion.

**Comment by iSazonov** [comment]
Yes, it males sense to add 1-2 tests for -like in Pester and full set of xUnit tests for Escape and UnEscape methods.
